src/goto-programs/remove_exceptions.cpp

```diff
@@ -0,0 +1,598 @@
+
+  // for now only add exceptional returns for Java
+  if(function_symbol.mode!=ID_java)
+    return;
+
+  if(goto_program.empty())
+    return;
+
+  // We generate an exceptional return value for any function that has
+  // a throw or a function call. This can be improved by only considering
+  // function calls that may escape exceptions. However, this will
+  // require multiple passes.
+  Forall_goto_program_instructions(instr_it, goto_program)
+  {
+    if(instr_it->is_throw() || instr_it->is_function_call())
+    {
+      // look up the function symbol
+      symbol_tablet::symbolst::iterator s_it=
+        symbol_table.symbols.find(function_id);
+
+      assert(s_it!=symbol_table.symbols.end());
+      const symbolt &function_symbol=s_it->second;
+
+      auxiliary_symbolt new_symbol;
+      new_symbol.is_static_lifetime=true;
+      new_symbol.module=function_symbol.module;
+      new_symbol.base_name=id2string(function_symbol.base_name)+EXC_SUFFIX;
+      new_symbol.name=id2string(function_symbol.name)+EXC_SUFFIX;
+      new_symbol.mode=function_symbol.mode;
+      new_symbol.type=typet(ID_pointer, empty_typet());
+      symbol_table.add(new_symbol);
+
+      // initialize the exceptional return with NULL
+      symbol_exprt lhs_expr_null;
+      lhs_expr_null.set_identifier(id2string(function_symbol.name)+EXC_SUFFIX);
+      lhs_expr_null.type()=typet(ID_pointer, empty_typet());
+
+      exprt rhs_expr_null;
+      rhs_expr_null=null_pointer_exprt(pointer_typet());
+      goto_programt::targett t_null=goto_program.insert_before(instr_it);
+      t_null->make_assignment();
+      t_null->source_location=instr_it->source_location;
+      t_null->code=code_assignt(
+        lhs_expr_null,
+        typecast_exprt(rhs_expr_null, lhs_expr_null.type()));
+      t_null->function=instr_it->function;
+      return;
+    }
+  }
+}
+
+/*******************************************************************\
+
+Function: remove_exceptionst::replace_throws
+
+Inputs:
+
+Outputs:
+
+Purpose: turns 'throw x' in function f into an assignment to f#exc_value
+
+\*******************************************************************/
+
+void remove_exceptionst::replace_throws(
+  const goto_functionst::function_mapt::iterator &func_it)
+{
+  const irep_idt &function_id=func_it->first;
+  goto_programt &goto_program=func_it->second.body;
+
+  if(goto_program.empty())
+    return;
+
+  Forall_goto_program_instructions(instr_it, goto_program)
+  {
+    if(instr_it->is_throw() &&
+       symbol_table.has_symbol(id2string(function_id)+EXC_SUFFIX))
+    {
+      assert(instr_it->code.operands().size()==1);
+      const symbolt &function_symbol=
+        symbol_table.lookup(id2string(function_id)+EXC_SUFFIX);
+
+      // replace "throw x;" by "f#exception_value=x;"
+      symbol_exprt lhs_expr;
+      lhs_expr.set_identifier(id2string(function_id)+EXC_SUFFIX);
+      lhs_expr.type()=function_symbol.type;
```

Comment:
This type seems different from the `void*` that is being generated in `add_exceptional_returns`?!